prow/flagutil/github.go

```diff
@@ -47,25 +47,62 @@ type GitHubOptions struct {
 	AppID             string
 	AppPrivateKeyPath string
 
+	ThrottleHourlyTokens int
+	ThrottleAllowBurst   int
+
 	// This will only be set after a github client was retrieved for the first time
 	appsTokenGenerator github.GitHubAppTokenGenerator
 }
 
 const DefaultGitHubTokenPath = "/etc/github/oauth" // Exported for testing purposes
 
+func githubDefaults() GitHubOptions {
+	return GitHubOptions{
+		Host:            github.DefaultHost,
+		endpoint:        NewStrings(github.DefaultAPIEndpoint),
+		graphqlEndpoint: github.DefaultGraphQLEndpoint,
+	}
+}
+
+type CustomGithubDefaulter func(options *GitHubOptions)
+
+func GithubThrottlerDefaults(hourlyTokens, allowedBursts int) CustomGithubDefaulter {
+	return func(o *GitHubOptions) {
+		o.ThrottleHourlyTokens = hourlyTokens
+		o.ThrottleAllowBurst = allowedBursts
+	}
+}
+
 // AddFlags injects GitHub options into the given FlagSet.
 func (o *GitHubOptions) AddFlags(fs *flag.FlagSet) {
-	fs.StringVar(&o.Host, "github-host", github.DefaultHost, "GitHub's default host (may differ for enterprise)")
-	o.endpoint = NewStrings(github.DefaultAPIEndpoint)
+	defaults := githubDefaults()
+	o.addFlags(fs, defaults)
+}
+
+// AddFlagsWithCustomDefaults injects GitHub options into the given FlagSet,
+func (o *GitHubOptions) AddFlagsWithCustomDefaults(fs *flag.FlagSet, defaulters ...CustomGithubDefaulter) {
```

Comment:
How do we deal with tide which constructs two clients with different defaults for their throttler settings? And also I would have thought we have more components with pre-existing flags for this

Reply:
> How do we deal with tide which constructs two clients with different defaults for their throttler settings?

The change should at least not break it, `Throttle` is smart enough and it can handle multiple subsequent `Throttle` calls, so Tide should just continue working same as it did before. The problematic part is probably that the components that do something smart like this will have confusing `--help` output and values passed in this new param would not be respected. Would that be a blocker?

> And also I would have thought we have more components with pre-existing flags for this

We probably do, but I haven't invested the effort in porting them all. I will do the legwork if the general approach is accepted.

Reply:
> The problematic part is probably that the components that do something smart like this will have confusing --help output and values passed in this new param would not be respected. Would that be a blocker?

Can we maybe just add an option to not add the new throttler flags if requested and use that?

> We probably do, but I haven't invested the effort in porting them all. I will do the legwork if the general approach is accepted.

sounds reasonable

Reply:
One way how to make Tide conform is to infer its partial throttlers from the totals given by the params. That way Tide will first create the clients trottled to the specified values (right now the clients are unthrottled) and then subsequent `Throttle` calls cal further restrict the individual clients.

Reply:
> Can we maybe just add an option to not add the new throttler flags if requested and use that?

Yep. It does not fit into the current pattern but it should not be hard to bend it for that purpose.